File: lib/languages/python/function-arguments.py
# Basic function arguments
def add(a, b):
   return a + b

# Default parameters
def greet(name, greeting="Hello"):
   return f"{greeting}, {name}!"

# Variable positional arguments
def sum_all(*numbers):
   return sum(numbers)

# Variable keyword arguments
def create_user(**user_data):
   return {
       "id": generate_id(),
       **user_data
   }

# Keyword-only arguments (Python 3+)
def configure(*, host="localhost", port=8080):
   return f"{host}:{port}"

# Positional-only arguments (Python 3.8+)
def divide(a, b, /):
   return a / b

# Combining argument types
def process(a, b, /, c, *, d, **kwargs):
   # a, b are positional-only
   # c is positional or keyword
   # d is keyword-only
   # kwargs captures other keyword arguments
   return a + b + c + d + sum(kwargs.values())

# Unpacking arguments
def point_distance(x1, y1, x2, y2):
   return ((x2 - x1) ** 2 + (y2 - y1) ** 2) ** 0.5

point1 = (1, 2)
point2 = (4, 6)
distance = point_distance(*point1, *point2)

# Unpacking keyword arguments
config = {"host": "example.com", "port": 443}
server = configure(**config)

# Type hints for complex arguments
from typing import List, Dict, Any, Optional

def analyze(
   data: List[Dict[str, Any]],
   fields: Optional[List[str]] = None,
   *,
   sort_by: str = "id"
) -> Dict[str, Any]:
   # Implementation
   return {"result": "analysis"}

def generate_id():
   import random
   return random.randint(1000, 9999)

File: lib/languages/elixir/function-arguments.ex
# Basic function arguments
def add(a, b) do
  a + b
end

# Default arguments
def greet(name, greeting \\ "Hello") do
  "#{greeting}, #{name}!"
end

# Pattern matching in function clauses
def process({:ok, value}), do: value
def process({:error, _}), do: 0

# Guards in function arguments
def abs(n) when n < 0, do: -n
def abs(n) when n >= 0, do: n

# Variable arguments with lists
def sum(numbers) do
  Enum.sum(numbers)
end

# Keyword arguments
def create_user(name, opts \\ []) do
  age = Keyword.get(opts, :age, 0)
  email = Keyword.get(opts, :email, "")
  %{id: generate_id(), name: name, age: age, email: email}
end

# Calling with keyword arguments
user = create_user("Alice", age: 30, email: "alice@example.com")

# Function arity matters
def duplicate(x), do: [x, x]
def duplicate(x, n), do: List.duplicate(x, n)

# Destructuring in arguments
def process_coords({x, y}), do: x + y

defp generate_id do
  :crypto.strong_rand_bytes(16) |> Base.encode16(case: :lower)
end

